From f2034325efad3af93192f61bf549cbaac6839e9f Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 3 Jun 2015 19:38:28 -0400
Subject: [PATCH] Expose CommandMap and allow Command name changing

Many plugins prefer to implement their own flexible command system,
and avoid using the plugin.yml for command registration. Currently
this requires Reflection to access the command map to dynamically
register commands.

This commit exposes the Command Map so that you may register a
command without defining it in plugin.yml. This is extremely useful
to plugins that have a lot of commands and also want to dynamically
register commands based on context (such as plugin being used by
multiple servers, but certain commands only for certain server types).

Additionally, Extend the Command class to allow changing the name.
This enables the ability to use Annotation based name selection,
which is not possible to look up in the constructor before calling
super(name);
---
 src/main/java/org/bukkit/Server.java          | 10 +++++++++-
 src/main/java/org/bukkit/command/Command.java | 18 +++++++++++++++++-
 2 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e1d89a5..87e49e9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -14,6 +14,7 @@ import java.util.logging.Logger;
 
 import org.bukkit.Warning.WarningState;
 import org.bukkit.command.CommandException;
+import org.bukkit.command.CommandMap;
 import org.bukkit.command.CommandSender;
 import org.bukkit.command.ConsoleCommandSender;
 import org.bukkit.command.PluginCommand;
@@ -358,11 +359,18 @@ public interface Server extends PluginMessageRecipient {
     /**
      * Gets a services manager.
      *
-     * @return s services manager
+     * @return a services manager
      */
     public ServicesManager getServicesManager();
 
     /**
+     * Returns the servers command map for managing commands.
+     *
+     * @return The servers command map
+     */
+    public CommandMap getCommandMap();
+
+    /**
      * Gets a list of all worlds on this server.
      *
      * @return a list of worlds
diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index c72c8fa..0bae9ba 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -21,7 +21,8 @@ import com.google.common.collect.ImmutableList;
  * Represents a Command, which executes various tasks upon user input
  */
 public abstract class Command {
-    private final String name;
+    private String name;
+    private String nextName;
     private String nextLabel;
     private String label;
     private List<String> aliases;
@@ -40,6 +41,7 @@ public abstract class Command {
 
     protected Command(String name, String description, String usageMessage, List<String> aliases) {
         this.name = name;
+        this.nextName = name;
         this.nextLabel = name;
         this.label = name;
         this.description = description;
@@ -119,6 +121,19 @@ public abstract class Command {
     }
 
     /**
+     * Sets the name of this command.
+     *
+     * @param name New command name
+     */
+    public void setName(String name) {
+        if (!isRegistered()) {
+            this.name = name;
+        } else {
+            this.nextName = name;
+        }
+    }
+
+    /**
      * Gets the permission required by users to be able to perform this
      * command
      *
@@ -245,6 +260,7 @@ public abstract class Command {
         if (allowChangesFrom(commandMap)) {
             this.commandMap = null;
             this.activeAliases = new ArrayList<String>(this.aliases);
+            this.name = this.nextName;
             this.label = this.nextLabel;
             return true;
         }
-- 
1.9.1

